### PR TITLE
Move targets, rzc and extension assembly in to the Sdk

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: src/Microsoft.NET.Sdk.Razor/Microsoft.NET.Sdk.Razor.csproj [hunk 1]
```diff
@@ -41,4 +34,64 @@
       <Link>Shared\Client.cs</Link>
     </Compile>
   </ItemGroup>
+
+  <Target Name="LayoutDependencies" BeforeTargets="Build" Condition="'$(IsInnerBuild)' != 'true'">
```
Comment (pranavkm) [inline src/Microsoft.NET.Sdk.Razor/Microsoft.NET.Sdk.Razor.csproj]: Weird MSBuild way to make things run a target AfterBuild

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    -->
+    <ProvideApplicationPartFactoryAttributeTypeName Condition="'$(ProvideApplicationPartFactoryAttributeTypeName)' == ''">Microsoft.AspNetCore.Mvc.ApplicationParts.CompiledRazorAssemblyApplicationPartFactory, Microsoft.AspNetCore.Mvc.Razor</ProvideApplicationPartFactoryAttributeTypeName>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <_TargetingNETCoreApp30OrLater Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' AND '$(_TargetFrameworkVersionWithoutV)' &gt; '2.9'">true</_TargetingNETCoreApp30OrLater>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(RazorLangVersion)'==''">
+    <RazorLangVersion Condition="'$(_TargetingNETCoreApp30OrLater)' == 'true'">2.1</RazorLangVersion>
```
Comment (pranavkm) [inline src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets]: We determine the `RazorLangVersion` and `RazorDefaultConfiguration` based on TFM. We'd need for a way to allow class libraries to specify this.

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets [hunk 3]
```diff
@@ -22,10 +22,15 @@ Copyright (c) .NET Foundation. All rights reserved.
...
+    <!-- Paths to tools, tasks, and extensions are calculated relative to the RazorSdkDirectoryRoot. This can be modified to test a local build. -->
+    <RazorSdkDirectoryRoot Condition="'$(RazorSdkDirectoryRoot)'==''">$(MSBuildThisFileDirectory)..\..\</RazorSdkDirectoryRoot>
+    <RazorSdkBuildTasksDirectoryRoot Condition="'$(RazorSdkBuildTasksDirectoryRoot)'==''">$(RazorSdkDirectoryRoot)tasks\</RazorSdkBuildTasksDirectoryRoot>
     <_RazorSdkTasksTFM Condition=" '$(MSBuildRuntimeType)' == 'Core'">netstandard2.0</_RazorSdkTasksTFM>
     <_RazorSdkTasksTFM Condition=" '$(_RazorSdkTasksTFM)' == ''">net46</_RazorSdkTasksTFM>
     <RazorSdkBuildTasksAssembly>$(RazorSdkBuildTasksDirectoryRoot)$(_RazorSdkTasksTFM)\Microsoft.NET.Sdk.Razor.Tasks.dll</RazorSdkBuildTasksAssembly>
+
+    <_TargetFrameworkVersionWithoutV>$(TargetFrameworkVersion.TrimStart('vV'))</_TargetFrameworkVersionWithoutV>
+    <_EnableAllInclusiveRazorSdk Condition="'$(_EnableInclusiveRazorSdk)' == '' AND '$(TargetFrameworkIdentifier)' == '.NETCoreApp' AND '$(_TargetFrameworkVersionWithoutV)' &gt; '2.9'">true</_EnableAllInclusiveRazorSdk>
```
Comment (pranavkm) [inline src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets]: Needs a name. The SDK pivots on this value to determine if it needs to use the compiler \ targets from the Sdk. A 3.0 RCL would need to be able to specify this.

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets [hunk 4]
```diff
@@ -123,6 +128,11 @@ Copyright (c) .NET Foundation. All rights reserved.
     <!-- Output directory used for generated files -->
     <RazorGenerateIntermediateOutputPath Condition="'$(RazorGenerateIntermediateOutputPath)'==''">$(IntermediateOutputPath)Razor\</RazorGenerateIntermediateOutputPath>
 
+    <!--
+      Use the suffix .Views when producing compiled view assemblies. This matches the requirements for Mvc's ViewsFeatureProvider.
+    -->
+    <RazorTargetNameSuffix Condition="'$(_EnableInclusiveRazorSdk)' == '' AND '$(RazorTargetNameSuffix)'==''">.Views</RazorTargetNameSuffix>
```
Comment (pranavkm) [inline src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets]: Need to play with this - ideally this is set based on the configuration. That said, I'm not super sure this is a particularly important problem to solve.

File: test/Microsoft.NET.Sdk.Razor.Test/IntegrationTests/BuildIntegrationTest.cs [hunk 5]
```diff
@@ -614,6 +614,22 @@ public async Task Build_WithDeterministicFlagSet_OutputsDeterministicViewsAssemb
             Assert.Equal(firstAssemblyBytes, secondAssemblyBytes);
         }
 
+        [Fact]
+        [InitializeTestProject("SimpleMvc21")]
+        public async Task Building_NETCoreApp21TargetingProject()
```
Comment (pranavkm) [inline test/Microsoft.NET.Sdk.Razor.Test/IntegrationTests/BuildIntegrationTest.cs]: Verifying that 2.1 projects build when using the new Sdk. Will add a scenario for 2.2 once it ships

File: test/Microsoft.NET.Sdk.Razor.Test/IntegrationTests/InitializeTestProjectAttribute.cs [hunk 6]
```diff
@@ -37,7 +37,7 @@ public override void Before(MethodInfo methodUnderTest)
             }
 
             MSBuildIntegrationTestBase.Project = ProjectDirectory.Create(_originalProjectName, _testProjectName, _baseDirectory, _additionalProjects, _language);
-            MSBuildIntegrationTestBase.TargetFramework = _originalProjectName.StartsWith("ClassLibrary") ? "netstandard2.0" : "netcoreapp2.2";
+            MSBuildIntegrationTestBase.TargetFramework = _originalProjectName.StartsWith("ClassLibrary") ? "netstandard2.0" : "netcoreapp3.0";
```
Comment (pranavkm) [inline test/Microsoft.NET.Sdk.Razor.Test/IntegrationTests/InitializeTestProjectAttribute.cs]: #yolo

File: test/testapps/AppWithP2PReference/AppWithP2PReference.csproj [hunk 7]
```diff
@@ -1,19 +1,11 @@
 <Project Sdk="Microsoft.NET.Sdk.Web">
   
   <PropertyGroup>
-    <_RazorMSBuildRoot>$(SolutionRoot)src\Microsoft.AspNetCore.Razor.Design\bin\$(Configuration)\</_RazorMSBuildRoot>
-    <RazorSdkBuildTasksDirectoryRoot>$(RazorSdkProjectDirectory)bin\$(Configuration)\</RazorSdkBuildTasksDirectoryRoot>
+    <RazorSdkDirectoryRoot>$(RazorSdkProjectDirectory)bin\$(Configuration)\sdk-output\</RazorSdkDirectoryRoot>
```
Comment (pranavkm) [inline test/testapps/AppWithP2PReference/AppWithP2PReference.csproj]: Building the sdk lays out binaries the same way it looks like in the package. Makes it easy to have one property to define all paths.

**Comment by rynowak** [comment]
👏 👏

File: NuGetPackageVerifier.json [hunk 8]
```diff
@@ -17,54 +17,53 @@
...
     ],
     "packages": {
-      "Microsoft.AspNetCore.Razor.Design": {
+      "Microsoft.NET.Sdk.Razor": {
         "exclusions": {
           "ASSEMBLY_DESCRIPTION": {
-            "tools/Microsoft.CodeAnalysis.CSharp.dll": "This assembly is not owned by us and does not follow our conventions.",
-            "tools/Microsoft.CodeAnalysis.dll": "This assembly is not owned by us and does not follow our conventions."
+            "tools/netcoreapp2.0/Microsoft.CodeAnalysis.CSharp.dll": "This assembly is not owned by us and does not follow our conventions.",
+            "tools/netcoreapp2.0/Microsoft.CodeAnalysis.dll": "This assembly is not owned by us and does not follow our conventions."
```
Comment (rynowak) [inline NuGetPackageVerifier.json]: Do we need to ship these here now that we are in the SDK? I'm not sure how all of this stuff gets constructed.

File: src/Microsoft.NET.Sdk.Razor/Microsoft.NET.Sdk.Razor.csproj [hunk 9]
```diff
@@ -1,30 +1,23 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <Description>Razor is a markup syntax for adding server-side logic to web pages. This package contains MSBuild support for Razor.</Description>
-    <TargetFrameworks>netstandard2.0;net46</TargetFrameworks>
+    <TargetFrameworks>netcoreapp2.2;netstandard2.0;net46</TargetFrameworks>
```
Comment (rynowak) [inline src/Microsoft.NET.Sdk.Razor/Microsoft.NET.Sdk.Razor.csproj]: hmmm?

File: src/Microsoft.NET.Sdk.Razor/Microsoft.NET.Sdk.Razor.csproj (hunk 1 above)
Comment (rynowak) [reply src/Microsoft.NET.Sdk.Razor/Microsoft.NET.Sdk.Razor.csproj]: What's wrong with `AfterTargets="AfterBuild"`?

File: src/Microsoft.NET.Sdk.Razor/Microsoft.NET.Sdk.Razor.csproj [hunk 10]
```diff
@@ -41,4 +34,64 @@
...
+    <SignedPackageFile Include="$(SdkOutputPath)tasks\netstandard2.0\$(TargetName).dll" Certificate="$(AssemblySigningCertName)"  />
+    <SignedPackageFile Include="$(SdkOutputPath)extensions\**\*.dll" Certificate="$(AssemblySigningCertName)"  />
+
+    <SignedPackageFile Include="$(SdkOutputPath)tools\netcoreapp2.0\rzc.dll" Certificate="$(AssemblySigningCertName)" />
+    <SignedPackageFile Include="$(SdkOutputPath)tools\netcoreapp2.0\Newtonsoft.Json.dll" Certificate="3PartyDual" />
+
+    <!-- Binaries that should be signed by corefx/roslyn -->
+    <ExcludePackageFileFromSigning Include="$(SdkOutputPath)tools\netcoreapp2.0\Microsoft.CodeAnalysis.CSharp.dll" />
+    <ExcludePackageFileFromSigning Include="$(SdkOutputPath)tools\netcoreapp2.0\Microsoft.CodeAnalysis.dll" />
+    <ExcludePackageFileFromSigning Include="$(SdkOutputPath)tools\netcoreapp2.0\runtimes\unix\lib\netstandard1.3\System.Text.Encoding.CodePages.dll" />
```
Comment (rynowak) [inline src/Microsoft.NET.Sdk.Razor/Microsoft.NET.Sdk.Razor.csproj]: This seems really fragile

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.CodeGeneration.targets [hunk 11]
```diff
@@ -14,27 +26,32 @@
...
 
-  <!-- 
+  <!--
     Consider these properties to be private to this targets file. The main Razor SDK should define all of the properties
-    that we use to pass data back and forth. 
+    that we use to pass data back and forth.
   -->
   <PropertyGroup>
+    <!-- Continue setting this property to maintain compat with legacy Razor.Design -->
+    <IsRazorCompilerReferenced>true</IsRazorCompilerReferenced>
```
Comment (rynowak) [inline src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.CodeGeneration.targets]: So if you are using the 3.0 SDK on a 2.0 project ..... how does this behave?

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+          impossible to load or build your projects from the command-line or the IDE.
+
+Copyright (c) .NET Foundation. All rights reserved.
+***********************************************************************************************
+-->
+<Project ToolsVersion="14.0">
+  <!--
+      Initialize properties and items inferred using the project version. This file is not imported in projects referencing
+      MVC \ Razor 2.2 or earlier since values specified here are provided via targets and props imported from NuGet packages
+      such as Microsoft.AspNetCore.Razor.Design.
```
Comment (rynowak) [inline src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets]: Should `IsRazorCompilerReferenced` be here?

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+      The properties and items here are designed to be read by CPS so they should be just simple evaluation-time values
+      and should not require targets to initialize.
+  -->
+  <PropertyGroup>
+    <!--
+      MVC uses a ProvideApplicationPartFactoryAttribute on the generated assembly to load compiled views from assembly. Set this to false, to prevent generating this attribute.
+    -->
+    <GenerateProvideApplicationPartFactoryAttribute>true</GenerateProvideApplicationPartFactoryAttribute>
```
Comment (rynowak) [inline src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets]: Should this be in a .props? can the user's project override this

Comment (rynowak) [reply src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets]: Or have a condition

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets]: 👍

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+  <PropertyGroup Condition="'$(RazorLangVersion)'==''">
+    <RazorLangVersion Condition="'$(_TargetingNETCoreApp30OrLater)' == 'true'">2.1</RazorLangVersion>
+  </PropertyGroup>
+
+  <!--
+    Set the primary configuration supported by this SDK as the default configuration for Razor.
+  -->
+  <PropertyGroup Condition="'$(RazorDefaultConfiguration)'==''">
+    <RazorDefaultConfiguration Condition="'$(_TargetingNETCoreApp30OrLater)' == 'true'">MVC-2.1</RazorDefaultConfiguration>
+  </PropertyGroup>
```
Comment (rynowak) [inline src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets]: FYI we have a work item logged to create a 3.0 version of all of these assets. This is fine for now.

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.DesignTime.targets [hunk 15]
```diff
@@ -21,6 +21,12 @@ Copyright (c) .NET Foundation. All rights reserved.
       we use the lack of the capability to detect downlevel scenarios.
     -->
     <ProjectCapability Include="DotNetCoreRazor"/>
+
+    <!--
+      Defines the ability to understand the configuration for the Razor language service provided by
+      the runtime/toolset packages. Introduced in 2.1
+    -->
+    <ProjectCapability Include="DotNetCoreRazorConfiguration"/>
```
Comment (rynowak) [inline src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.DesignTime.targets]: Should this be conditional? How would this work for a project targeting 2.0 but using a new VS/new SDK?

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets (hunk 3 above)
Comment (rynowak) [reply src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets]: `_EnableAllInclusiveRazorSdk` is dope.

Can you log a work item for 3.0 RCL support? I don't think we should try to solve that in this PR

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets (hunk 4 above)
Comment (rynowak) [reply src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets]: yeah that seems fine for now.

File: test/Microsoft.NET.Sdk.Razor.Test/IntegrationTests/BuildIntegrationTest.cs (hunk 5 above)
Comment (rynowak) [reply test/Microsoft.NET.Sdk.Razor.Test/IntegrationTests/BuildIntegrationTest.cs]: Log an issue :)

File: test/Microsoft.NET.Sdk.Razor.Test/IntegrationTests/InitializeTestProjectAttribute.cs (hunk 6 above)
Comment (rynowak) [reply test/Microsoft.NET.Sdk.Razor.Test/IntegrationTests/InitializeTestProjectAttribute.cs]: Yeah, this is fine for now. We should log an issue for 3.0 RCL support and solve this then.

File: test/testapps/AppWithP2PReference/AppWithP2PReference.csproj (hunk 7 above)
Comment (rynowak) [reply test/testapps/AppWithP2PReference/AppWithP2PReference.csproj]: 👍

File: test/testapps/ClassLibrary/ClassLibrary.csproj [hunk 16]
```diff
@@ -1,25 +1,20 @@
...
-
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
 
     <Copyright>© Microsoft</Copyright>
     <Product>Razor Test</Product>
     <Company>Microsoft</Company>
     <Description>ClassLibrary Description</Description>
+
+    <_EnableAllInclusiveRazorSdk>true</_EnableAllInclusiveRazorSdk>
```
Comment (rynowak) [inline test/testapps/ClassLibrary/ClassLibrary.csproj]: hmmmm - I think I want to understand this

File: NuGetPackageVerifier.json (hunk 8 above)
Comment (pranavkm) [reply NuGetPackageVerifier.json]: Filed https://github.com/aspnet/Razor/issues/2675 to figure this out.

File: src/Microsoft.NET.Sdk.Razor/Microsoft.NET.Sdk.Razor.csproj (hunk 9 above)
Comment (pranavkm) [reply src/Microsoft.NET.Sdk.Razor/Microsoft.NET.Sdk.Razor.csproj]: Required to reference Razor.Tools. Once we reference it we can make the build order more reliable and avoid file contention.

File: src/Microsoft.NET.Sdk.Razor/Microsoft.NET.Sdk.Razor.csproj (hunk 1 above)
Comment (pranavkm) [reply src/Microsoft.NET.Sdk.Razor/Microsoft.NET.Sdk.Razor.csproj]: Apparently this is the way to run something after build - https://github.com/Microsoft/msbuild/issues/1680#issuecomment-297450755

File: src/Microsoft.NET.Sdk.Razor/Microsoft.NET.Sdk.Razor.csproj (hunk 10 above)
Comment (pranavkm) [reply src/Microsoft.NET.Sdk.Razor/Microsoft.NET.Sdk.Razor.csproj]: I agree. Maybe addressing https://github.com/aspnet/Razor/issues/2675 would fix this issue. That said, it's been copied over from the Razor.Design project so this isn't new:

https://github.com/aspnet/Razor/blob/master/src/Microsoft.AspNetCore.Razor.Design/Microsoft.AspNetCore.Razor.Design.csproj#L39-L42

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.CodeGeneration.targets (hunk 11 above)
Comment (pranavkm) [reply src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.CodeGeneration.targets]: This file only gets imported if a property is set (set by default for netcoreapp 3.0 or later projects).

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets (hunk 12 above)
Comment (pranavkm) [reply src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets]: Sure

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets (hunk 13 above)
Comment (pranavkm) [reply src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Configuration.targets]: Condition seems better. We want to set these up in 3.0 projects, a props file is too early for checking this.

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.DesignTime.targets (hunk 15 above)
Comment (pranavkm) [reply src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.DesignTime.targets]: Yup. I can move this to `.Configuration.targets`

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets (hunk 3 above)
Comment (pranavkm) [reply src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets]: https://github.com/aspnet/Razor/issues/2676

File: test/Microsoft.NET.Sdk.Razor.Test/IntegrationTests/BuildIntegrationTest.cs (hunk 5 above)
Comment (pranavkm) [reply test/Microsoft.NET.Sdk.Razor.Test/IntegrationTests/BuildIntegrationTest.cs]: https://github.com/aspnet/Razor/issues/2677

File: test/testapps/ClassLibrary/ClassLibrary.csproj (hunk 16 above)
Comment (pranavkm) [reply test/testapps/ClassLibrary/ClassLibrary.csproj]: These are *technically* 3.0 targeting RCLs (without references to Razor.Design). Consequently we need for a way to tell the SDK to use the targets it ships. Essentially this - https://github.com/aspnet/Razor/issues/2676. There's an existing 2.1 RCL testapp that does not require this

**Pull request merged**